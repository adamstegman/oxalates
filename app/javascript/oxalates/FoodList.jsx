import React from 'react';
import PropTypes from 'prop-types';

import { foodPropType } from './foodPropType';
import { FoodListEmptyItem } from './FoodListEmptyItem';
import { FoodListError } from './FoodListError';
import { FoodListItem } from './FoodListItem';
import { FoodListLoading } from './FoodListLoading';
import { FoodListNewItem } from './FoodListNewItem';
import { listPropType } from './listPropType';

import './FoodList.scss';

export class FoodList extends React.Component {
  // Find the "highest" (most dangerous) list
  getListForFood(food, lists) {
    return lists.find(list => (
      (
        (list.bottomThreshold && list.bottomThreshold <= food.oxalateMg) &&
        (list.topThreshold && list.topThreshold >= food.oxalateMg)
      ) ||
      (list.bottomThreshold === null && (list.topThreshold && list.topThreshold >= food.oxalateMg)) ||
      ((list.bottomThreshold && list.bottomThreshold <= food.oxalateMg) && list.topThreshold === null)
    ));
  }

  sortFoodsAlphabetically(foods) {
    return foods.sort((a, b) => a.name.localeCompare(b.name));
  }

  sortListsByThresholds(lists) {
    return lists.sort((a, b) => {
      const aBottom = a.bottomThreshold || 0;
      const aTop = a.topThreshold || Number.MAX_VALUE;
      const aTotal = aBottom + aTop;
      const bBottom = b.bottomThreshold || 0;
      const bTop = b.topThreshold || Number.MAX_VALUE;
      const bTotal = bBottom + bTop;
      if (aTotal < bTotal) {
        return 1;
      } else if (aTotal > bTotal) {
        return -1;
      } else {
        return 0;
      }
    });
  }

  render() {
    const {
      foods,
      lists,
      requestedListId,
      newFood,
      newFoodListId,
      error,
      password,
      createFood,
      setNewFood,
    } = this.props;
    const sortedFoods = this.sortFoodsAlphabetically(foods);
    const sortedLists = this.sortListsByThresholds(lists);

    let contents = <FoodListEmptyItem />;
    if (requestedListId) {
      contents = <FoodListLoading />;
    } else if (newFood) {
      contents = <FoodListNewItem newFood={newFood}
                                  newFoodListId={newFoodListId}
                                  createFood={createFood}
                                  setNewFood={setNewFood}
                                  error={error}
                                  password={password} />;
    } else if (error) {
      contents = <FoodListError error={error} />;
    } else if (foods.length > 0) {
      contents = sortedFoods.map(food => {
        const list = this.getListForFood(food, sortedLists);
        return <FoodListItem key={food.id} food={food} list={list} />;
      });
    }

    return (
      <div className="content">
        <ul className="food-list">
          {contents}
        </ul>
      </div>
    );
  }
};

FoodList.propTypes = {
  foods: PropTypes.arrayOf(foodPropType).isRequired,
  lists: PropTypes.arrayOf(listPropType).isRequired,
  requestedListId: PropTypes.node,
  error: PropTypes.string,
  newFood: PropTypes.shape({}),
  newFoodListId: PropTypes.node,
  password: PropTypes.string,
  createFood: PropTypes.func,
  setNewFood: PropTypes.func,
};
