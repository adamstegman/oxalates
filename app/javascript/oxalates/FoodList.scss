$food_list_item_border: 1px solid #ddd;

.food-list {
  list-style-type: none;
  margin: 0;
  padding: 0;

  & > li {
    display: -webkit-flex;
    display: flex;
    -webkit-align-items: center;
    align-items: center;

    border-bottom: $food_list_item_border;
  }
}
