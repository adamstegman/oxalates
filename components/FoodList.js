import React from 'react';
import {
  FlatList,
  StyleSheet,
  View,
} from 'react-native';

import Colors from '../constants/Colors';
import { FoodListItem } from './FoodListItem';
import { FoodService } from '../services/FoodService';

export class FoodList extends React.Component {
  constructor() {
    super();
    this.state = { foods: [] };
    this.foodService = new FoodService();
  }

  componentDidMount() {
    let bottomThreshold = undefined;
    let topThreshold = undefined;
    if (this.props.list) {
      bottomThreshold = this.props.list.bottomThreshold;
      topThreshold = this.props.list.topThreshold;
    }
    this.foodService.query({ bottomThreshold, topThreshold }).then((foods) => {
      this.setState({ foods });
    });
  }

  separator = () => {
    return (
      <View
        style={{
          height: 1,
          backgroundColor: Colors.listSeparator,
        }}
      />
    );
  };

  render() {
    return (
      <View style={styles.container}>
        <FlatList data={this.state.foods}
                  renderItem={({ item }) => <FoodListItem food={item} />}
                  keyExtractor={(food, index) => index}
                  ItemSeparatorComponent={this.separator}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.backgroundColor,
  },
});
