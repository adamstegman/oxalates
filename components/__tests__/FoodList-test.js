import 'react-native';
import React from 'react';
import { FoodList } from '../FoodList';
import renderer from 'react-test-renderer';

it('renders a list of foods', () => {
  const tree = renderer.create(<FoodList list={{name: 'High', bottomThreshold: 15, topThreshold: 50, color: '#e00'}} />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders a list of all foods', () => {
  const tree = renderer.create(<FoodList />).toJSON();

  expect(tree).toMatchSnapshot();
});
