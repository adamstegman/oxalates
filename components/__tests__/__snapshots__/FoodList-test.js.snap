// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a list of all foods 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;

exports[`renders a list of foods 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
